src/main/kotlin/org/partiql/scribe/targets/spark/SparkCalls.kt

```diff
@@ -119,30 +115,83 @@ public open class SparkCalls(private val log: ProblemCallback) : SqlCalls() {
     /**
      * PartiQL's date_diff accepts a part whereas Spark's date_diff returns the difference in days between two dates.
      *
-     * We extract the datetime part and perform the difference.
+     * For date parts, we use the builtins (or calculate years)
+     * For time parts, we convert to unix seconds and convert using 60 seconds per minute and 60 minutes per hour.
      *
      * Notes:
      *  > https://github.com/partiql/partiql-lang-kotlin/wiki/Functions#date_diff----since-v010
      *  > https://spark.apache.org/docs/latest/api/java/org/apache/spark/sql/functions.html#datediff(org.apache.spark.sql.Column,org.apache.spark.sql.Column)
      *  > https://spark.apache.org/docs/2.3.0/api/sql/index.html#datediff
      */
-    override fun dateDiff(part: DatetimeField, args: SqlArgs): Expr {
-        val extract = when (part) {
-            DatetimeField.YEAR -> "year"
-            DatetimeField.MONTH -> "month"
-            DatetimeField.DAY -> "day"
-            DatetimeField.HOUR -> "hour"
-            DatetimeField.MINUTE -> "minute"
-            DatetimeField.SECOND -> "second"
-            else -> error("Unexpected datetime part `$part`")
+    override fun dateDiff(part: DatetimeField, args: SqlArgs): Expr = when (part) {
+        DatetimeField.YEAR -> {
+            val d1 = args[0].expr
+            val d2 = args[1].expr
+            val call = exprCall(
+                function = id("months_between"),
+                args = listOf(d2, d1)
+            )
+            log.info("PartiQL `date_diff(year, <date_1>, <date_2>)` was replaced by Spark `floor(months_between(<date_2>, <date_1>) / 12).")
```

Comment:
I'm not sure if `floor` is what we want for `date_diff` (months, years, hour, and minute). Consider if first datetime is earlier than the second datetime, the `months_between`/`diff` will return a negative value.

E.g.: date_diff between two timestamps that are 1 minute and 1 second apart

```
PartiQL> DATE_DIFF(minute, `1997-02-28T10:30:01Z`, `1997-02-28T10:29:00Z`);
==='
-1
```

spark will output something different

```
spark-sql (default)> SELECT floor((unix_timestamp('1997-02-28 10:29:00') - unix_timestamp('1997-02-28 10:30:01')) / 60);
-2
```

Probably what we need is some kind of decimal truncation or rounding to 0?

Reply:
Good call. `CAST(.. AS INT)` will truncate.

Reply:
```sql
spark-sql (default)> SELECT CAST((`unix_timestamp`(TIMESTAMP'1997-02-28T10:29:00Z') - `unix_timestamp`(TIMESTAMP'1997-02-28T10:30:01Z')) / 60 AS BIGINT);
-1
```